docs/proposals/auth.md

```diff
@@ -0,0 +1,150 @@
+Heapster Authentication and Authorization
+=========================================
+
+Current State
+-------------
+
+Heapster authentication is based on checking client certificates against a CA.
+Authorization is either just a static list of names, or non-existant.  In
+either case, anyone authenticated is authorized to perform any operation.  With
+push metrics, the proposed plan was to allow adding a second CA, to allow
+separating authorization for model and Oldtimer queries from authorization to
+push metrics.  However, this is suboptimal, since it also separates
+authentication.
+
+Requirements
+------------
+
+Heapster should be able to separate authentication from authorization, and
+the authorization interface should be flexible enough to support authorization
+implementations that are able to make the following distinctions:
+
+- containing queries to a particular namespace: we should be able to only allow
```

Comment:
> The alternative is to run a heapster per namespace. This solves both authz and qos problems. If you have a free tier where you can't afford to run heapster per namespace, you can use a shared instance, but restrict access to it.

I think that heapster per-namespace is going to have similar requirements about separating readers and writers.  The kube `TokenAccessReview`, `SubjectAccessReview` and Impersonation features were intended to enable use-cases like this one.  I think this will be good experience for us as we try to move those APIs out of the beta stage.

Also, this feature wouldn't preclude per-namespace heapster scenarios if we want to try them out.